### PR TITLE
add keycloak to k8s for ingress nginx

### PR DESCRIPTION
**Pull request by indiepopart**
Add Keycloak prod deployment to k8s sub-generator for ingress nginx edge option, including PostgreSQL database and clustering. HTTPS/TLS is not configured. This allows testing on minikube.

---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [ ] The JDL part is updated if necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [x] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

File: generators/kubernetes/templates/README-KUBERNETES.md.ejs [hunk 1]
```diff
@@ -162,6 +167,10 @@ Events:
 
 cert-manager first creates a temporary certificate. Once you see the event "The certificate has been successfully issued", it means Let's Encrypt staging has issued the certificate. The staging certificate will be trusted by deployed applications, but not by browsers. Let's Encrypt production certificates are trusted by browsers, but the production issuer will probably not work with nip.io, and fail with "too many certificates already issued for nip.io".
 
+<<<<<<< HEAD
+  <%_ } _%>
+=======
+>>>>>>> upstream/main
```
Comment (mraible) [inline generators/kubernetes/templates/README-KUBERNETES.md.ejs:173]: It looks like a merge issue needs to be resolved here.

File: generators/kubernetes/templates/README-KUBERNETES.md.ejs [hunk 2]
```diff
@@ -179,11 +188,19 @@ This can occur if your cluster has low resource (e.g. Minikube). Increase the `i
...
 
 The k8s sub-generator configures Let's Encrypt staging environment, which is not trusted by browsers and Java applications by default. In the Kubernetes descriptors, the staging CAs are added to the applications truststore and registry truststore. Note that cert-manager issues a temporary certificate first, but the applications and registry don't trust the ad-hoc CA, and will fail the startup until the Let's Encrypt certificate is ready and updated in the ingress service. You will see the application pod STATUS as Error and CrashLoopBackOff during the multiple startup attempts. You can check the certificate status with the following command:
 
 ```
 kubectl describe certificate keycloak-ssl <%= kubernetesNamespaceDefault ? '' : ` -n ${kubernetesNamespace}` %>
 ```
+<<<<<<< HEAD
+  <%_ } _%>
+=======
+>>>>>>> upstream/main
```
[the remark this answers is not in the document]
Comment (indiepopart) [reply generators/kubernetes/templates/README-KUBERNETES.md.ejs]: Fixed

File: generators/kubernetes/templates/README-KUBERNETES.md.ejs [hunk 3]
```diff
@@ -179,11 +188,19 @@ This can occur if your cluster has low resource (e.g. Minikube). Increase the `i
 The default setting are optimized for middle-scale clusters. You are free to increase the JAVA_OPTS environment variable, and resource requests and limits to improve the performance. Be careful!
 
 <%_ if (useKeycloak) { _%>
+<<<<<<< HEAD
+  <%_ if (ingressTypeGke) { _%>
+=======
+>>>>>>> upstream/main
```
[the remark this answers is not in the document]
Comment (indiepopart) [reply generators/kubernetes/templates/README-KUBERNETES.md.ejs]: Fixed

File: generators/kubernetes/templates/README-KUBERNETES.md.ejs (hunk 1 above)
Comment (indiepopart) [reply generators/kubernetes/templates/README-KUBERNETES.md.ejs]: Fixed

File: generators/kubernetes/templates/deployment.yml.ejs [hunk 4]
```diff
@@ -216,12 +216,19 @@ spec:
...
           value: "x-request-id,x-ot-span-context"
 <%_ } _%>
-<%_ if (useKeycloak) { _%>  
+<%_ if (useKeycloak) { _%>
+      <%_ if (ingressTypeGke) { _%>
         # Custom trustStore required when using Let's Encrypt staging
         - name: JAVA_OPTS
-          value: "-Djavax.net.ssl.trustStore=/etc/pki/java/cacerts -Djavax.net.ssl.trustStorePassword=123456 -Xmx256m -Xms256m"
-        - name: SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_OIDC_ISSUER_URI        
+          value: " -Djavax.net.ssl.trustStore=/etc/pki/java/cacerts -Djavax.net.ssl.trustStorePassword=123456 -Xmx256m -Xms256m"
```
Comment (mraible) [inline generators/kubernetes/templates/deployment.yml.ejs:223]:
```suggestion
          value: "-Djavax.net.ssl.trustStore=/etc/pki/java/cacerts -Djavax.net.ssl.trustStorePassword=123456 -Xmx256m -Xms256m"
```

File: generators/kubernetes/templates/deployment.yml.ejs [hunk 5]
```diff
@@ -216,12 +216,19 @@ spec:
...
+        - name: SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_OIDC_ISSUER_URI
           value: https://keycloak.<%= kubernetesNamespace %>.<%= ingressDomain %>/realms/jhipster
+      <%_ } else { _%>
+        - name: JAVA_OPTS
+          value: " -Xmx256m -Xms256m"
```
Comment (mraible) [inline generators/kubernetes/templates/deployment.yml.ejs:228]:
```suggestion
          value: "-Xmx256m -Xms256m"
```

File: generators/kubernetes/templates/registry/jhipster-registry.yml.ejs [hunk 6]
```diff
@@ -126,11 +126,16 @@ spec:
         - name: K8S_CONFIG_PATH
           value: '/central-config/'
   <%_ if (useKeycloak) { _%>
+        <%_ if (ingressTypeGke) { _%>
         - name: SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_OIDC_ISSUER_URI
           value: https://keycloak.<%= kubernetesNamespace %>.<%= ingressDomain %>/realms/jhipster
         # Custom trustStore required when using Let's Encrypt staging
         - name: JAVA_OPTS
-          value: "-Djavax.net.ssl.trustStore=/etc/pki/java/cacerts -Djavax.net.ssl.trustStorePassword=123456"
+          value: " -Djavax.net.ssl.trustStore=/etc/pki/java/cacerts -Djavax.net.ssl.trustStorePassword=123456"
```
Comment (mraible) [inline generators/kubernetes/templates/registry/jhipster-registry.yml.ejs:134]:
```suggestion
          value: "-Djavax.net.ssl.trustStore=/etc/pki/java/cacerts -Djavax.net.ssl.trustStorePassword=123456"
```

**Comment by mraible** [comment]
I've been testing this for the last couple of days and can't get it to work on macOS. I was able to get everything to start on my M1, but I had to build M1-compatible images and deploy those to Docker Hub. 

We should change the commands that are output to the console for deploying directly since they don't currently contain `-PjibArchitecture=arm64`. For example:

```
./gradlew -Pprod bootJar jib -PjibArchitecture=arm64 -Djib.to.image=mraible/gateway
```

I've been talking with @indiepopart and ideally, this should work after installing this branch:

First, start Minikube and enable ingress:

```
minikube start --cpus 6
minikube addons enable ingress
minikube ip
```

You will need to use the output of `minikube ip` + `.nip.io` when prompted. For example, `192.168.58.2.nip.io`. 

Then, create a microservices stack and run the `k8s` sub-generator:

```
take reactive-stack
jhipster jdl reactive-mf
# Deploy images to Docker Hub
rm -rf kubernetes
take k8s
jhipster k8s # choose Ingress and NGINX
./kubectl-apply.sh -f
```

**Comment by mraible** [comment]
@indiepopart I think we need to document a couple of things for this PR and Minikube: 

1. That you need to run `minikube addons enable ingress`. 
2. That it won't work on macOS. https://github.com/kubernetes/minikube/issues/13951

The best place to add documentation for Minikube is probably https://www.jhipster.tech/kubernetes/. However, I'm not sure people will see it there, so outputting it to the console might be a good idea too.

**Comment by indiepopart** [comment]
I woud rather change the generator and documentation so it works for also for macOS or Windows. Avoid using the minikube ip and nip.io.

**Comment by indiepopart** [comment]
@mraible The README was updated with instructions to use `minikube tunnel` for exposing the ingress controller to the host. The deployment was only verified on Linux.

**Review by mraible — DISMISSED** [review]

**Review by mraible — APPROVED** [review]

**Pull request merged**